### ISSUE TITLE
Fragment caching doesn't work in mailer views

### ISSUE THREAD

**Issue by rwz** [rails/rails#17657]
Rails 4.1.8

Steps to reproduce: 

```sh
$ rails new cache_fail
$ cd cache_fail
$ rails g mailer foo_mailer bar
$ edit app/views/foo_mailer/bar.html.erb
```

Edit file to contain this:

```erb
<%- cache "foo" do %>
bar
<%- end %>
```

Fire up the console and execute:

```
> FooMailer.bar.deliver
```

Get this error:

```
FooMailer#bar: processed outbound mail in 169.2ms
ActionView::Template::Error: undefined method `perform_caching' for #<FooMailer:0x007f88974f3568>
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/helpers/cache_helper.rb:114:in `cache'
        from /Users/rwz/dev/cache_fail/app/views/foo_mailer/bar.html.erb:3:in `_app_views_foo_mailer_bar_html_erb__3101049261217162251_70112315263580'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/template.rb:145:in `block in render'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications.rb:161:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/template.rb:339:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/template.rb:143:in `render'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/template_renderer.rb:55:in `block (2 levels) in render_template'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/abstract_renderer.rb:38:in `block in instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications.rb:159:in `block in instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications.rb:159:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/abstract_renderer.rb:38:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/template_renderer.rb:54:in `block in render_template'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/template_renderer.rb:62:in `render_with_layout'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/template_renderer.rb:53:in `render_template'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionview-4.1.8/lib/action_view/renderer/template_renderer.rb:17:in `render'
... 17 levels...
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionmailer-4.1.8/lib/action_mailer/base.rb:580:in `block in process'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications.rb:159:in `block in instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/activesupport-4.1.8/lib/active_support/notifications.rb:159:in `instrument'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionmailer-4.1.8/lib/action_mailer/base.rb:577:in `process'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionmailer-4.1.8/lib/action_mailer/base.rb:568:in `initialize'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionmailer-4.1.8/lib/action_mailer/base.rb:551:in `new'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/actionmailer-4.1.8/lib/action_mailer/base.rb:551:in `method_missing'
        from (irb):1
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/railties-4.1.8/lib/rails/commands/console.rb:90:in `start'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/railties-4.1.8/lib/rails/commands/console.rb:9:in `start'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/railties-4.1.8/lib/rails/commands/commands_tasks.rb:69:in `console'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/railties-4.1.8/lib/rails/commands/commands_tasks.rb:40:in `run_command!'
        from /Users/rwz/.rvm/gems/ruby-2.1.5@cache-fail/gems/railties-4.1.8/lib/rails/commands.rb:17:in `<top (required)>'
        from /Users/rwz/dev/cache_fail/bin/rails:4:in `require'
        from /Users/rwz/dev/cache_fail/bin/rails:4:in `<main>'2.1.5 :002 >
```

**Comment by viniciusnz** [comment]
+1 @rwz and @raja . @wdperson , we use it heavily and it has cut our sending times to a fraction of what they used to be, due to the repetition of different items for each person in lists, each item with lots of information and db queries. It used to work on Rails 3.2 and doesn't work any more in 4.2 due to the new fragment caching strategy

In rails 4.2 I also got this error `undefined local variable or method controller_name for XxxxMailer` so I had to also add to your monkey-patch the controller_name.

File `config/initializers/base_email.rb` looks like:

```
class ActionMailer::Base
  def perform_caching
    Rails.configuration.action_controller.perform_caching
  end
  def controller_name
    "ActionMailer" # anything, really
  end
end
```
Important note:

I also had to add `skip_digest: true` to the end of my cache calls like, for example:
```
 <% cache([item], :skip_digest => true) do %>
``` 

Obviously this is not an end solution but got my cache to work again.
